File: go/cose/encrypt0_test.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

package cose

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// TestEncrypt0Vector checks the COSE_Encrypt0 transform against the Python-
// generated vector: a fixed IV reproduces the sealed bytes, and the frozen
// COSE opens back to the plaintext.
func TestEncrypt0Vector(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join("..", "..", "vectors", "encrypt0", "basic.json"))
	if err != nil {
		t.Fatalf("vectors/encrypt0/basic.json must exist: %v", err)
	}
	var c struct {
		Key       string `json:"key"`
		IV        string `json:"iv"`
		Kid       string `json:"kid"`
		Plaintext string `json:"plaintext"`
		Cose      string `json:"cose"`
	}
	if err := json.Unmarshal(raw, &c); err != nil {
		t.Fatal(err)
	}
	key, _ := hex.DecodeString(c.Key)
	iv, _ := hex.DecodeString(c.IV)
	plaintext, _ := hex.DecodeString(c.Plaintext)
	expected, _ := hex.DecodeString(c.Cose)

	// Fixed IV -> the sealed bytes reproduce the frozen vector exactly.
	sealed, err := Encrypt0WithIV(plaintext, c.Kid, key, iv)
	if err != nil {
		t.Fatalf("seal: %v", err)
	}
	if !bytes.Equal(sealed, expected) {
		t.Errorf("sealed mismatch:\n got %x\nwant %x", sealed, expected)
	}

	// The recipient kid round-trips out of the cleartext header.
	if kid, ok := RecipientKID(expected); !ok || kid != c.Kid {
		t.Errorf("RecipientKID = %q, %v; want %q", kid, ok, c.Kid)
	}

	// The frozen COSE opens back to the plaintext under the content key.
	got, err := Decrypt0(expected, func(kid string) ([]byte, bool) {
		return key, kid == c.Kid
	})
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	if !bytes.Equal(got, plaintext) {
		t.Errorf("opened = %x, want %x", got, plaintext)
	}

	// No key -> ErrMissingKey; wrong key -> ErrAuthFailed.
	if _, err := Decrypt0(expected, func(string) ([]byte, bool) { return nil, false }); err != ErrMissingKey {
		t.Errorf("missing key err = %v, want %v", err, ErrMissingKey)
	}
	wrong := make([]byte, 32)
	if _, err := Decrypt0(expected, func(string) ([]byte, bool) { return wrong, true }); err != ErrAuthFailed {
		t.Errorf("wrong key err = %v, want %v", err, ErrAuthFailed)
	}
}

func TestEncrypt0RandomIVRoundTrip(t *testing.T) {
	key := make([]byte, 32)
	for i := range key {
		key[i] = byte(i)
	}
	sealed, err := Encrypt0([]byte("verified id record"), "did:court", key)
	if err != nil {
		t.Fatalf("seal: %v", err)
	}
	got, err := Decrypt0(sealed, func(kid string) ([]byte, bool) {
		return key, kid == "did:court"
	})
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	if string(got) != "verified id record" {
		t.Errorf("round-trip = %q", got)
	}
}

File: ts/tests/encrypt0.test.ts
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// COSE_Encrypt0 conformance (§9.3): the TypeScript engine reproduces the frozen
// fixed-IV AES-256-GCM seal (vectors/encrypt0/basic.json) byte-for-byte, opens
// it, and round-trips a random-IV seal.

import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "node:test";
import assert from "node:assert/strict";

import {
    encrypt0,
    encrypt0WithIv,
    decrypt0,
    recipientKid,
    Encrypt0Error,
} from "../src/cose.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "../../../");

function unhex(s: string): Uint8Array {
    return Uint8Array.from(Buffer.from(s, "hex"));
}

test("encrypt0 vector: fixed-IV seal reproduces frozen bytes and opens (§9.3)", () => {
    const c = JSON.parse(
        readFileSync(join(repoRoot, "vectors", "encrypt0", "basic.json"), "utf-8"),
    );
    const key = unhex(c.key);
    const iv = unhex(c.iv);
    const plaintext = unhex(c.plaintext);
    const expected = unhex(c.cose);

    // Fixed IV -> the sealed bytes reproduce the frozen vector exactly.
    assert.deepEqual(encrypt0WithIv(plaintext, c.kid, key, iv), expected);

    // The recipient kid round-trips out of the cleartext header.
    assert.equal(recipientKid(expected), c.kid);

    // The frozen COSE opens back to the plaintext under the content key.
    assert.deepEqual(
        decrypt0(expected, (kid) => (kid === c.kid ? key : null)),
        plaintext,
    );

    // No key -> missing-key; wrong key -> auth-failed.
    assert.throws(
        () => decrypt0(expected, () => null),
        (e: unknown) => e instanceof Encrypt0Error && e.reason === "missing-key",
    );
    assert.throws(
        () => decrypt0(expected, () => new Uint8Array(32)),
        (e: unknown) => e instanceof Encrypt0Error && e.reason === "auth-failed",
    );
});

test("encrypt0 random-IV round-trip", () => {
    const key = new Uint8Array(32).map((_, i) => i);
    const sealed = encrypt0(new TextEncoder().encode("verified id record"), "did:court", key);
    assert.equal(recipientKid(sealed), "did:court");
    assert.deepEqual(
        decrypt0(sealed, (kid) => (kid === "did:court" ? key : null)),
        new TextEncoder().encode("verified id record"),
    );
});
